### DISCUSSION TITLE
openssl 1.0.2k Unable to upgrade openssl 1.1.1v

### DISCUSSION THREAD

**Discussion by geekops-cn** [openssl/openssl#21776] [Q&A] [answered]
## System Related Information:

- CentOS Linux release 7.9.2009 (Core)
- Kernel: Linux 3.10.0-1160.95.1.el7.x86_64
- OpenSSL 1.0.2k-fips 26 Jan 2017

I am currently running a Nessus scan on my system which has detected vulnerabilities in OpenSSL. It requires upgrading OpenSSL 1.0.2k to 1.1.1v on this CentOS system. However, I'm encountering issues during the installation process and unable to perform the upgrade.

#### Questions:

1. Does CentOS 7 support the installation of OpenSSL 1.1.1v?
2. What is the correct procedure for installing OpenSSL 1.1.1v?
3. After successfully installing OpenSSL 1.1.1v and verifying with `openssl version`, the system reports OpenSSL 1.1.1v. However, Nessus continues to detect the following information:
   - Path             : /usr/lib64/libcrypto.so.1.0.2k
   - Reported version : 1.0.2k
   - Fixed version    : 1.0.2zg

I'm seeking guidance on how to properly address these issues and perform the required upgrade. Any assistance would be greatly appreciated. Thank you.

**Comment by t8m** [comment, marked as answer]
In general questions about the system openssl packages in various OSes should be directed at the vendors of those OSes.

But to answer what I can answer.

> 1. Does CentOS 7 support the installation of OpenSSL 1.1.1v?

Not as the system package. It should be possible to install 1.1.1v in addition to the system 1.0.2k package.

> 2. What is the correct procedure for installing OpenSSL 1.1.1v?

Normal configure & build process should work. You must not specify prefix in a way that the resulting installation would overwrite the system libcrypto/libssl shared libraries as that would break your system.

>     3. After successfully installing OpenSSL 1.1.1v and verifying with `openssl version`, the system reports OpenSSL 1.1.1v. However, Nessus continues to detect the following information:
>        
>        * Path             : /usr/lib64/libcrypto.so.1.0.2k
>        * Reported version : 1.0.2k
>        * Fixed version    : 1.0.2zg

The reason is that it looks at the openssl library coming from the system openssl package which cannot be changed to 1.1.1v.

Any software that uses OpenSSL that needs to be using 1.1.1v needs to be built against the 1.1.1v version.

To summarize, unless you really know what you're doing and there is only one or two services that you need to be using the current OpenSSL version that you can rebuild yourself, either use properly supported OS version from a vendor or upgrade to a never version of CentOS.

**Reply by geekops-cn** [reply]
**Thank you very much for your response.** I understand you to mean that in CentOS 7, OpenSSL 1.0.2k is the default system version and cannot be modified or upgraded. I can only separately deploy OpenSSL 1.1.1v. Is that correct?

**My current confusion is that the security team is asking me to address OpenSSL vulnerabilities "OpenSSL 1.0.2 < 1.0.2zg Multiple Vulnerabilities"** based on the Nessus report. How should I resolve the OpenSSL issue?

**I have already installed OpenSSL 1.1.1v using** `openssl-1.1.1v.tar.gz` **and verified it with** `openssl version` **displaying "OpenSSL 1.1.1v 1 Aug 2023".** However, after running the Nessus scan again, it still prompts for an OpenSSL upgrade. This leaves me puzzled and unsure about what to do.

![image](https://github.com/openssl/openssl/assets/50939257/2f83a32e-98ca-4a11-9b4c-d88923f58f8e)

**Reply by t8m** [reply]
Unfortunately I cannot help you further with this. I do not know what exactly the Nessus scan scans. And yes, the system 1.0.2k openssl really cannot be replaced.

**Reply by t8m** [reply]
If the scan looks directly at the /usr/lib64/libcrypto.so.1.0.2k file then this cannot be fixed.

**Comment by tom-cosgrove-arm** [comment]
> in CentOS 7, OpenSSL 1.0.2k is the default system version and cannot be modified or upgraded. I can only separately deploy OpenSSL 1.1.1v. Is that correct?

My reading would be:you have OpenSSL 1.0.2k installed on your CentOS 7 system. You can't change this to something like OpenSSL 1.1.1 as the API is incompatible, but you can upgrade to later versions of 1.0.2. _However_ OpenSSL 1.0.2 is out of support since 1st January 2020 and is no longer receiving updates (although extended support is available from OpenSSL Software Services for premium support customers).

This sounds like something you should be discussing with your security team

**Reply by geekops-cn** [reply]
Thank you for your response. The default OpenSSL version in `CentOS Linux release 7.9.2009 (Core)` is `1.0.2k`. The latest security team scan report has identified vulnerabilities in `OpenSSL 1.0.2k` and recommends an upgrade. However, `CentOS Linux release 7.9.2009 (Core)` no longer supports an upgrade to `OpenSSL 1.0.2k`. Even after installing `OpenSSL 1.1.1v` on the system, the security team can still detect that `1.0.2k` has not been upgraded and remains vulnerable. Therefore, the current solution is to switch to a different operating system such as Ubuntu or SUSE.

**Reply by paulidale** [reply]
An alternative would be to take out a premium support contract to gain access to the latest 1.0.2 sources, and then build and install that.

**Reply by geekops-cn** [reply]
Thank you very much for your reply. Based on the current Centos7 version, we will not choose premium subscription to fix this issue. There is currently only the option to stop using centos7. Switch to ubuntu or opensuse.

Don't know other centos7 used to face the openssl 1.0.2 k security problem how to deal with?

**Reply by paulidale** [reply]
Without taking out a support contract, you need to switch to a new operating system.
There is nothing else we can suggest.
